File: go_server/communication/communication.go
package communication

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"sync"
	"taller_apirest/models"

	"github.com/nats-io/nats.go"
)

var (
	once     sync.Once
	instance *nats.Conn
	logger   *NatsLogger // Variable global para acceder a la instancia única de NatsLogger
)

// NatsLogger es una estructura que contiene la conexión a NATS
type NatsLogger struct {
	conn *nats.Conn
}

func ConnectToNATS() *NatsLogger {
	var natsHost string
	natsHost = os.Getenv("NATS_SERVER")
	if natsHost == "" {
		natsHost = "localhost"
	}
	url := "nats://" + natsHost + ":4222"
	once.Do(func() {
		nc, err := nats.Connect(url)
		if err != nil {
			log.Fatalf("Error al conectar con NATS: %v", err)
		}
		instance = nc
	})

	return &NatsLogger{conn: instance}
}

func init() {
	logger = ConnectToNATS()
}

// SendLog envía un mensaje al tema de NATS usando la información en Notification
func (nl *NatsLogger) SendLog(newLog *models.LogResponse) {

	var subject string

	subject = os.Getenv("NATS_SUBJECT")
	if subject == "" {
		subject = "MicroservicesLogs"
	}

	// Convertir la estructura en JSON
	jsonData, err := json.Marshal(newLog)
	if err != nil {
		log.Fatal(err)
	}
	// return nl.conn.Publish(subject, []byte(notification.Message))
	// Publicar el mensaje JSON
	if err := nl.conn.Publish("MicroservicesLogs", jsonData); err != nil {
		log.Fatal(err)
	}

	fmt.Println("Notification sent")
}


func (nl *NatsLogger) NotifyUserRegistration(name, email, log_type string) bool{
	var subject string = "users.creation"
	msg := models.Registration{
		Name:  name,
		Email: email,
		Type:  log_type,
	}
	jsonData, err := json.Marshal(msg)
	if err != nil {
		return false
	}
	if err := nl.conn.Publish(subject, jsonData); err != nil {
		return false
	}
	return true

}

func (nl *NatsLogger) SendSampleMessage() bool {
	var subject string = "test"

	if err := nl.conn.Publish(subject, []byte("Sample message")); err != nil {
		return false
	}

	return true
}

// HealthCheckNATS verifica si la conexión a NATS está viva
func (nl *NatsLogger) HealthCheckNATS() bool {
	if nl.conn == nil {
		return false
	} else if nl.conn.Status() != nats.CONNECTED {
		return false
	}
	return true
}

// ReadyNats verifica si se puede enviar un mensaje por NATS
func (nl *NatsLogger) ReadyNats() bool {

	return nl.SendSampleMessage()

}

File: notification/communication/communication.py
from nats.aio.client import Client as NATS
from models.log import Log
import asyncio
import datetime
import os

class NATSClient:
    _instance = None
    _is_connected = False

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(NATSClient, cls).__new__(cls)
            cls._instance.nc = NATS()
        return cls._instance

    async def connect(self):
        if not self._is_connected:
            nats_host = os.getenv("NATS_SERVER", "localhost")
            nats_url = "nats://" + nats_host + ":4222"
            try:
                await self.nc.connect(nats_url)
                self._is_connected = True
            except Exception as e:
                print(f"Failed to connect to NATS: {e}")
                self._is_connected = False

    async def publish(self, subject, message):
        await self.nc.publish(subject, message)

    async def drain(self):
        await self.nc.drain()

    async def close(self):
        await self.nc.close()
        self._is_connected = False

async def send_log_via_nats(log):
    nats_client = NATSClient()
    await nats_client.connect()

    if nats_client._is_connected:
        log_json = log.to_json()
        await nats_client.publish("MicroservicesLogs", log_json.encode('utf-8'))

def create_log(name, summary, description, log_type):
    log = Log(
        name=name,
        summary=summary,
        description=description,
        log_type=log_type,
        log_date=datetime.datetime.now(),
        module="NOTIFICATION-API"
    )
    asyncio.run(send_log_via_nats(log))

async def test_connection():
    nats_client = NATSClient()
    await nats_client.connect()
    return nats_client._is_connected

async def send_sample_message():
    nats_client = NATSClient()
    await nats_client.connect()

    if nats_client._is_connected:
        log = Log(
            name="Test",
            summary="Test",
            description="Test",
            log_type="INFO",
            log_date=datetime.datetime.now(),
            module="NOTIFICATION-API"
        )
        log_json = log.to_json()
        await nats_client.publish("sample", log_json.encode('utf-8'))
        await nats_client.drain()
        await nats_client.close()